res/controllers/lib/midi-controls-0.js

```diff
@@ -0,0 +1,651 @@
+    }
+
+    if (typeof this.unshift === 'function') {
+        this.unshift();
+    }
+    // These cannot be in the prototype; they must be unique to each instance.
+    this.isShifted = false;
+    this.connections = [];
+
+    if (options !== undefined && typeof options.co === 'string') {
+        this.inCo = options.co;
+        this.outCo = options.co;
+    }
+
+    if (this.outConnect && this.group !== undefined && this.outCo !== undefined) {
+        this.connect();
+        if (this.outTrigger) {
+            this.trigger();
+        }
+    }
+};
+Control.prototype = {
+    // default attributes
+    // You should probably overwrite at least some of these.
+    inValueScale: function (value) {
+        return value / this.max;
+    },
+    // map input in the XML file, not inValueScale
+    input: function (channel, control, value, status, group) {
+        this.setParameter(this.inValueScale(value));
+    },
+    max: 127, // for MIDI. When adapting for HID this may change.
+    outValueScale: function (value) {return value * this.max;},
+    output: function (value, group, control) {
+        this.send(this.outValueScale(value));
+    },
+    outConnect: true,
+    outTrigger: true,
+
+    // common functions
+    // In most cases, you should not overwrite these.
+    setValue: function (value) {
+        engine.setValue(this.group, this.inCo, value);
+    },
+    setParameter: function (value) {
+        print (this.inCo);
+        engine.setParameter(this.group, this.inCo, value);
+    },
+    // outCo value generally shouldn't be set directly,
+    // only by the output() callback when its value changes,
+    // or by calling trigger()
+    // so don't provide separate setValueIn/setValueOut functions.
+    getValueIn: function () {
+        return engine.getValue(this.group, this.inCo);
+    },
+    getValueOut: function () {
+        return engine.getValue(this.group, this.outCo);
+    },
+    getParameterIn: function () {
+        return engine.getParameter(this.group, this.inCo);
+    },
+    getParameterOut: function () {
+        return engine.getParameter(this.group, this.outCo);
+    },
+    toggle: function () {
+        this.setValue( ! this.getValueIn());
+    },
+    connect: function () {
+        /**
+        Override this method with a custom one to connect multiple Mixxx COs for a single Control.
+        Add the connection objects to the this.connections array so they all get disconnected just
+        by calling this.disconnect(). This can be helpful for multicolor LEDs that show a
+        different color depending on the state of different Mixxx COs. See SamplerButton.connect()
+        and SamplerButton.output() for an example.
+        **/
+        if (undefined !== this.group &&
+            undefined !== this.outCo &&
+            undefined !== this.output &&
+            typeof this.output === 'function') {
+            this.connections[0] = engine.connectControl(this.group, this.outCo, this.output);
```

Comment:
However, these connection objects should have a `trigger()` method. Currently the only way to trigger them is with `engine.trigger(group, co)`, and I don't know how that behaves if multiple callbacks are made to the same group & CO.